tools/doc-generator/doc-generator.go

```diff
@@ -0,0 +1,103 @@
+package main
+
+import (
+	"bufio"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"sort"
+	"strings"
+
+	fake "kubevirt.io/kubevirt/pkg/monitoring/vms/prometheus"
+	promvm "kubevirt.io/kubevirt/pkg/monitoring/vms/prometheus" // import for prometheus metrics
+	_ "kubevirt.io/kubevirt/pkg/virt-controller/watch"
+)
+
+func main() {
+	req, err := http.NewRequest(http.MethodGet, "/metrics", nil)
+	checkError(err)
+
+	recorder := httptest.NewRecorder()
+
+	handler := promvm.Handler(1)
+
+	fake.RegisterFakeCollector()
+
+	handler.ServeHTTP(recorder, req)
+
+	if status := recorder.Code; status == http.StatusOK {
+		metrics := parseVirtMetrics(recorder.Body, []string{"kubevirt_", "_virt_controller"})
+		writeToFile(metrics)
+	} else {
+		panic(recorder.Code)
+	}
+}
+
+type promOutput struct {
+	Help string
+}
+
+func writeToFile(metrics map[string]promOutput) {
+	old, err := os.Open("metrics.md")
+	checkError(err)
+	defer old.Close()
+
+	new, err := os.Create("newmetrics.md")
+	checkError(err)
+	defer new.Close()
+
+	buf := bufio.NewWriter(new)
+	scan := bufio.NewScanner(old)
+	for scan.Scan() {
+		line := scan.Text()
+		buf.WriteString(line + "\n")
+		for k, v := range metrics {
+			if strings.Contains(line, k) {
+				buf.WriteString("###" + v.Help + "\n")
+				delete(metrics, k)
+				break
+			}
+		}
+	}
+
+	if len(metrics) > 0 {
+		buf.WriteString("\n # Other Metrics \n")
+	}
+
+	var keys []string
+	for k := range metrics {
+		keys = append(keys, k)
+	}
+
+	sort.Strings(keys)
+	for _, k := range keys {
+		buf.WriteString("## " + k + "\n")
+		buf.WriteString("###" + metrics[k].Help + "\n")
+	}
+	buf.Flush()
+}
+
+func parseVirtMetrics(r io.Reader, contains []string) map[string]promOutput {
```

Comment:
Regarding readability, I'd move this function before `writeToFile` as this one is called first. So the reader does not have to scroll up/down that much ;-)